File: src/shared/cloze.ts
/**
 * Cloze deletion parsing — Anki syntax: {{c1::answer}} or {{c1::answer::hint}}.
 * One card is generated per distinct cloze index; rendering blanks only the
 * active index and shows every other cloze's answer plainly (Anki behavior).
 */

const CLOZE_RE = /\{\{c(\d+)::([\s\S]*?)(?:::([\s\S]*?))?\}\}/g;

export interface ClozeSegment {
  text: string;
  /** Present when this segment is a cloze deletion */
  cloze?: { index: number; active: boolean };
}

/** Distinct cloze indexes in the text, sorted ascending. [] = not a valid cloze note. */
export function clozeIndexes(text: string): number[] {
  const indexes = new Set<number>();
  for (const match of text.matchAll(CLOZE_RE)) {
    const index = Number(match[1]);
    if (index >= 1) indexes.add(index);
  }
  return [...indexes].sort((a, b) => a - b);
}

/**
 * Split cloze text into renderable segments for one card (activeIndex).
 * Front: active clozes become "[...]" (or "[hint]"); others show their answer.
 * Back: active clozes show their answer (marked active so the UI can accent it).
 */
export function renderCloze(
  text: string,
  activeIndex: number,
  side: 'front' | 'back',
): ClozeSegment[] {
  const segments: ClozeSegment[] = [];
  let last = 0;
  for (const match of text.matchAll(CLOZE_RE)) {
    const index = Number(match[1]);
    const answer = match[2];
    const hint = match[3];
    const start = match.index;
    if (start > last) segments.push({ text: text.slice(last, start) });
    if (index < 1) {
      // Malformed index (c0): leave the raw text in place
      segments.push({ text: match[0] });
    } else if (index === activeIndex) {
      const blank = side === 'front' ? `[${hint ?? '...'}]` : answer;
      segments.push({ text: blank, cloze: { index, active: true } });
    } else {
      segments.push({ text: answer, cloze: { index, active: false } });
    }
    last = start + match[0].length;
  }
  if (last < text.length) segments.push({ text: text.slice(last) });
  return segments;
}

/** Convenience for plain-text previews (browse table, tests) */
export function clozeText(text: string, activeIndex: number, side: 'front' | 'back'): string {
  return renderCloze(text, activeIndex, side)
    .map((s) => s.text)
    .join('');
}

File: ios/ADHDReaderCore/Sources/ADHDReaderCore/Cloze.swift
import Foundation

/// Cloze deletion parsing — Anki syntax {{c1::answer}} or {{c1::answer::hint}}.
/// Ported from `src/shared/cloze.ts`.
public enum Cloze {
    // {{c(\d+)::(answer)(?:::(hint))?}}
    static let regex = try! NSRegularExpression(
        pattern: #"\{\{c(\d+)::([\s\S]*?)(?:::([\s\S]*?))?\}\}"#)

    private static func matches(in text: String) -> [NSTextCheckingResult] {
        let range = NSRange(text.startIndex..., in: text)
        return regex.matches(in: text, range: range)
    }

    private static func group(_ match: NSTextCheckingResult, _ i: Int, in text: String) -> String? {
        let r = match.range(at: i)
        guard r.location != NSNotFound, let range = Range(r, in: text) else { return nil }
        return String(text[range])
    }

    /// Distinct cloze indexes in the text, sorted ascending. [] = not a valid cloze note.
    public static func clozeIndexes(_ text: String) -> [Int] {
        var indexes = Set<Int>()
        for m in matches(in: text) {
            if let s = group(m, 1, in: text), let index = Int(s), index >= 1 { indexes.insert(index) }
        }
        return indexes.sorted()
    }

    public struct Segment: Equatable, Sendable {
        public let text: String
        public let clozeIndex: Int?
        public let clozeActive: Bool?
    }

    /// Split cloze text into renderable segments for one card (`activeIndex`).
    /// `side` is "front" or "back".
    public static func renderCloze(_ text: String, activeIndex: Int, side: String) -> [Segment] {
        var segments: [Segment] = []
        var lastUTF16 = 0
        let ns = text as NSString
        for m in matches(in: text) {
            let start = m.range.location
            if start > lastUTF16 {
                segments.append(Segment(text: ns.substring(with: NSRange(location: lastUTF16, length: start - lastUTF16)),
                                        clozeIndex: nil, clozeActive: nil))
            }
            let indexStr = group(m, 1, in: text) ?? ""
            let index = Int(indexStr) ?? 0
            let answer = group(m, 2, in: text) ?? ""
            let hint = group(m, 3, in: text)
            if index < 1 {
                segments.append(Segment(text: ns.substring(with: m.range), clozeIndex: nil, clozeActive: nil))
            } else if index == activeIndex {
                let blank = side == "front" ? "[\(hint ?? "...")]" : answer
                segments.append(Segment(text: blank, clozeIndex: index, clozeActive: true))
            } else {
                segments.append(Segment(text: answer, clozeIndex: index, clozeActive: false))
            }
            lastUTF16 = m.range.location + m.range.length
        }
        if lastUTF16 < ns.length {
            segments.append(Segment(text: ns.substring(from: lastUTF16), clozeIndex: nil, clozeActive: nil))
        }
        return segments
    }

    /// Plain-text preview for one card.
    public static func clozeText(_ text: String, activeIndex: Int, side: String) -> String {
        renderCloze(text, activeIndex: activeIndex, side: side).map(\.text).joined()
    }
}
